[#8779] AccountAttributes: Remove HTML sanitization before saving for googleId, name, institute

ChooJeremy opened the pull request:
Fixes #8779

**Outline of Solution**

<!-- Tell us how you solved the issue. -->
<!-- If there are things you want the reviewers to focus on, include them here as well. -->
<!-- This portion can be skipped if the fix is trivial. -->
Basically just removed the sanitization in AccountAttributes and updated tests. I have tested with single quotes in the email (as per #8774) and it works.

I'm not sure if I need to write a migration script, let me know if I need to. Based on my checks, if there was something to be migrated, it wouldn't have made it to the database anyway because the application would have just rejected it.

Ready for review.

xpdavid commented on src/main/java/teammates/common/datatransfer/attributes/AccountAttributes.java:
```diff
@@ -181,9 +181,9 @@ public String getJsonString() {
 
     @Override
     public void sanitizeForSaving() {
-        this.googleId = SanitizationHelper.sanitizeForHtml(googleId);
```
Should use `SanitizationHelper.sanitizeGoogleId`?

xpdavid commented on src/main/java/teammates/common/datatransfer/attributes/AccountAttributes.java:
```diff
@@ -181,9 +181,9 @@ public String getJsonString() {
     @Override
     public void sanitizeForSaving() {
-        this.googleId = SanitizationHelper.sanitizeForHtml(googleId);
-        this.name = SanitizationHelper.sanitizeForHtml(name);
```
`SanitizationHelper.sanitizeName`?

xpdavid commented on src/main/java/teammates/common/datatransfer/attributes/AccountAttributes.java:
```diff
@@ -181,9 +181,9 @@ public String getJsonString() {
     public void sanitizeForSaving() {
-        this.googleId = SanitizationHelper.sanitizeForHtml(googleId);
-        this.name = SanitizationHelper.sanitizeForHtml(name);
-        this.institute = SanitizationHelper.sanitizeForHtml(institute);
```
`SanitizationHelper.sanitizeName` maybe?

ChooJeremy commented:
Done, ready for review.

I used SanitizationHelper.sanitizeTitle for institute sanitization (based on AccountAttributes's Builder build() method.)

xpdavid commented on package.json:
```diff
@@ -44,7 +44,7 @@
   },
   "devDependencies": {
     "@angular-builders/jest": "~7.1.2",
-    "@angular-devkit/build-angular": "~0.11.4",
```
Not relevant change?

xpdavid commented on src/test/java/teammates/test/cases/datatransfer/AccountAttributesTest.java:
```diff
@@ -95,9 +94,9 @@ public void testSanitizeForSaving() {
         AccountAttributes expectedAccount = createAccountAttributesToSanitize();
         actualAccount.sanitizeForSaving();
 
-        assertEquals(SanitizationHelper.sanitizeForHtml(expectedAccount.googleId), actualAccount.googleId);
```
So your tests should also reflect the changes?

e.g. Having some values that need to be sanitised and test whether the `actualAccount.sanitizeForSaving();` work or not?

ChooJeremy commented:
Sorry, forgot about the tests. Interestingly, this means that the tests would have passed no matter what, because the Builder in AccountAttributes sanitized the results when building the object.

I've changed the email test to add the single quote and added whitespace around all the values since the sanitization code is effectively just trimming the values.

Ready for review.

xpdavid commented on src/main/java/teammates/common/datatransfer/attributes/AccountAttributes.java:
```diff
@@ -92,6 +92,10 @@ public AccountAttributes build() {
             return accountAttributes;
         }
 
+        public AccountAttributes buildWithoutSanitizing() {
```
I don't think it is a good practice to introduce this method in production code just for testing.

xpdavid commented on src/test/java/teammates/test/cases/datatransfer/AccountAttributesTest.java:
```diff
@@ -95,9 +95,10 @@ public void testSanitizeForSaving() {
-        assertEquals(SanitizationHelper.sanitizeForHtml(expectedAccount.name), actualAccount.name);
-        assertEquals(SanitizationHelper.sanitizeForHtml(expectedAccount.institute), actualAccount.institute);
+        assertEquals(SanitizationHelper.sanitizeGoogleId(expectedAccount.googleId), actualAccount.googleId);
+        assertEquals(SanitizationHelper.sanitizeName(expectedAccount.name), actualAccount.name);
```
>  Interestingly, this means that the tests would have passed no matter what, because the Builder in AccountAttributes sanitized the results when building the object.

If that is the case. Maybe we can just do `assertEquals(expectedAccount.name, actualAccount.name)`

xpdavid replied:
Cool. We can do that.

ChooJeremy commented:
Ok, ready for review.

xpdavid commented on src/test/java/teammates/test/cases/datatransfer/AccountAttributesTest.java:
```diff
@@ -238,15 +239,14 @@ private AccountAttributes createValidAccountAttributesObject() {
+        unsanitizedAttributes.googleId = "    google'Id@gmail.com\t";
+        unsanitizedAttributes.name = "'name'\n\n";
+        unsanitizedAttributes.institute = "\\\t\n/";
+        unsanitizedAttributes.email = "&<email>&\n";
```
I think we should change this to a valid email address.

xpdavid commented on src/test/java/teammates/test/cases/datatransfer/AccountAttributesTest.java:
```diff
@@ -238,15 +239,14 @@ private AccountAttributes createValidAccountAttributesObject() {
-
+        AccountAttributes unsanitizedAttributes = AccountAttributes.builder().build();
+        unsanitizedAttributes.googleId = "    google'Id@gmail.com\t";
+        unsanitizedAttributes.name = "'name'\n\n";
```
There should be some extra space inside name in order to make sure `sanitizeName` is doing the job? `    'name'\n\n`?

xpdavid commented on src/test/java/teammates/test/cases/datatransfer/AccountAttributesTest.java:
```diff
@@ -238,15 +239,14 @@ private AccountAttributes createValidAccountAttributesObject() {
+        AccountAttributes unsanitizedAttributes = AccountAttributes.builder().build();
+        unsanitizedAttributes.googleId = "    google'Id@gmail.com\t";
+        unsanitizedAttributes.name = "'name'\n\n";
+        unsanitizedAttributes.institute = "\\\t\n/";
```
Same for institute.

ChooJeremy commented:
Ok, test data modified, whitespace added to the middle of the strings.

Ready for review.
